### ISSUE TITLE
Implement AV1125: Don't expose stateful objects through static members

### ISSUE THREAD

**Issue by bkoelman** [bkoelman/CSharpGuidelinesAnalyzer#5]
This analyzer should report on static property getters whose method body contains an execution path that returns a type that has non-static members.

**Comment by bkoelman** [comment]
After some more thinking, I came to the conclusion that the *only* reason to return an object from a static getter would be that it actually *has* state (non-static members). Otherwise the caller could directly access the target object (and the static getter could be omitted). 

Consider the next code snippet:

```csharp
class Program
{
    static void Main()
    {
        var result = C.D.F;
    }        
} 
    
public class C
{
    public static D D { get; } // violation of AV1125
}

public class D
{
    public int F;
}
```

In the snippet above, instance field `D.F` is accessed via `C.D` in `Main`. The declaration of `C.D` violates AV1125. To solve that, we can make field `D.F` static. In that case, `Main` can directly access the field and the indirection through `C` can be removed:

```csharp
class Program
{
    static void Main()
    {
        var result = D.F;
    }        
} 
    
public class C
{
    public static D D { get; } // unused
}

public class D
{
    public static int F;
}
```

Following this pattern leads to either making everything static or making nothing static.

Note that the analyzers for AV1008 (AvoidStaticClassesAnalyzer) and AV1745 (StaticClassesShouldOnlyContainExtensionMethodsAnalyzer) report declarations of static classes (except for extension methods), so going static all the way is not the solution here.

Another way of looking at this, is to report *usages* of static fields that return an object with state. In other words: we would report a warning on the invocation in `Main` instead of `C.D`. The `HttpContext` example falls in this category: its definition cannot be changed, but reporting accessing it could be helpful.

**Comment by bkoelman** [comment]
Private fields declared in types in assembly references [are not exposed through Roslyn APIs](https://github.com/dotnet/roslyn/issues/5049). For example, private fields inside `HttpContext` are inaccessible, so this class would not be considered stateful.

Conclusion: *Defining* state in static types is already reported by the analyzers for other rules. *Consuming* them from external libraries is undetectable. 

I don't see any useful scenario's left to report for this rule.
